File: src/mcp_toolkit/schema.py
"""Derive JSON Schema from a function signature.

Agents pick tools by reading their schemas, so the schema is part of the
prompt whether you think of it that way or not. Writing it by hand means it
drifts from the code; generating it from type hints and the docstring means
it cannot.
"""

from __future__ import annotations

import enum
import inspect
import re
import types
import typing
from collections.abc import Callable
from dataclasses import MISSING, fields, is_dataclass
from typing import Any

_PRIMITIVES: dict[Any, dict[str, Any]] = {
    str: {"type": "string"},
    int: {"type": "integer"},
    float: {"type": "number"},
    bool: {"type": "boolean"},
    type(None): {"type": "null"},
    Any: {},
}

_ARGS_HEADER = re.compile(r"^\s*(Args|Arguments|Parameters)\s*:\s*$", re.IGNORECASE)
_GOOGLE_PARAM = re.compile(r"^\s*(\*{0,2}\w+)\s*(?:\([^)]*\))?\s*:\s*(.+)$")
_SPHINX_PARAM = re.compile(r"^\s*:param\s+(?:[\w\[\], ]+\s+)?(\w+)\s*:\s*(.+)$")


def summarize_docstring(func: Callable[..., Any]) -> str:
    """Return everything above the ``Args:`` block, collapsed to one blob."""
    doc = inspect.getdoc(func) or ""
    lines: list[str] = []
    for line in doc.splitlines():
        if _ARGS_HEADER.match(line) or _SPHINX_PARAM.match(line):
            break
        lines.append(line)
    return "\n".join(lines).strip()


def parse_param_docs(func: Callable[..., Any]) -> dict[str, str]:
    """Pull per-parameter descriptions out of a Google- or Sphinx-style docstring."""
    doc = inspect.getdoc(func) or ""
    found: dict[str, str] = {}
    in_args = False
    current: str | None = None

    for raw in doc.splitlines():
        sphinx = _SPHINX_PARAM.match(raw)
        if sphinx:
            found[sphinx.group(1)] = sphinx.group(2).strip()
            current = sphinx.group(1)
            continue

        if _ARGS_HEADER.match(raw):
            in_args = True
            current = None
            continue

        if not in_args:
            continue

        if raw.strip() and not raw.startswith((" ", "\t")):
            # Dedented back to a new section such as ``Returns:``.
            break

        match = _GOOGLE_PARAM.match(raw)
        if match:
            current = match.group(1).lstrip("*")
            found[current] = match.group(2).strip()
        elif current and raw.strip():
            found[current] += " " + raw.strip()

    return found


def _unwrap_optional(annotation: Any) -> tuple[Any, bool]:
    """Reduce ``T | None`` to ``(T, True)``; leave everything else alone."""
    origin = typing.get_origin(annotation)
    if origin not in (typing.Union, types.UnionType):
        return annotation, False
    args = [a for a in typing.get_args(annotation) if a is not type(None)]
    if len(args) == 1:
        return args[0], True
    reduced = args[0]
    for extra in args[1:]:
        reduced = reduced | extra
    return reduced, True


def type_to_schema(annotation: Any) -> dict[str, Any]:
    """Translate a type annotation into a JSON Schema fragment."""
    if annotation is inspect.Parameter.empty:
        return {}

    annotation, _ = _unwrap_optional(annotation)

    if annotation in _PRIMITIVES:
        return dict(_PRIMITIVES[annotation])

    origin = typing.get_origin(annotation)
    args = typing.get_args(annotation)

    if origin is typing.Literal:
        schema: dict[str, Any] = {"enum": list(args)}
        kinds = {type(a) for a in args}
        if kinds == {str}:
            schema["type"] = "string"
        elif kinds <= {int, bool}:
            schema["type"] = "integer"
        return schema

    if origin in (list, set, frozenset, tuple):
        item = type_to_schema(args[0]) if args else {}
        return {"type": "array", "items": item}

    if origin is dict:
        value = type_to_schema(args[1]) if len(args) == 2 else {}
        return {"type": "object", "additionalProperties": value or True}

    if origin in (typing.Union, types.UnionType):
        return {"anyOf": [type_to_schema(a) for a in args]}

    if isinstance(annotation, type) and issubclass(annotation, enum.Enum):
        return {"enum": [member.value for member in annotation]}

    if is_dataclass(annotation):
        props: dict[str, Any] = {}
        required: list[str] = []
        for f in fields(annotation):
            props[f.name] = type_to_schema(f.type)
            if f.default is MISSING and f.default_factory is MISSING:  # type: ignore[misc]
                required.append(f.name)
        out: dict[str, Any] = {"type": "object", "properties": props}
        if required:
            out["required"] = required
        return out

    # Unknown types stay permissive rather than silently rejecting valid input.
    return {}


def schema_from_signature(func: Callable[..., Any]) -> dict[str, Any]:
    """Build the ``inputSchema`` object MCP expects for a tool.

    ``self``, ``cls`` and ``*args``/``**kwargs`` are skipped. Parameters with
    defaults become optional; everything else lands in ``required``.
    """
    try:
        hints = typing.get_type_hints(func)
    except Exception:  # pragma: no cover - exotic forward refs
        hints = {}

    signature = inspect.signature(func)
    docs = parse_param_docs(func)

    properties: dict[str, Any] = {}
    required: list[str] = []

    for name, param in signature.parameters.items():
        if name in ("self", "cls"):
            continue
        if param.kind in (param.VAR_POSITIONAL, param.VAR_KEYWORD):
            continue

        annotation = hints.get(name, param.annotation)
        prop = type_to_schema(annotation)
        if name in docs:
            prop["description"] = docs[name]
        if param.default is not inspect.Parameter.empty:
            # ``default: null`` tells the model nothing it cannot infer from the
            # parameter being optional, so leave it out.
            if param.default is not None and _is_json_scalar(param.default):
                prop["default"] = param.default
        else:
            required.append(name)
        properties[name] = prop

    schema: dict[str, Any] = {
        "type": "object",
        "properties": properties,
        "additionalProperties": False,
    }
    if required:
        schema["required"] = required
    return schema


def _is_json_scalar(value: Any) -> bool:
    return isinstance(value, (str, int, float, bool, type(None)))


def validate_arguments(schema: dict[str, Any], arguments: dict[str, Any]) -> list[str]:
    """Check *arguments* against *schema*, returning human-readable problems.

    Deliberately small: presence, primitive type, enum membership and unknown
    keys. That catches the mistakes models actually make. Reach for
    ``jsonschema`` if you need the full specification.
    """
    problems: list[str] = []
    properties: dict[str, Any] = schema.get("properties", {})

    for name in schema.get("required", []):
        if name not in arguments:
            problems.append(f"missing required argument {name!r}")

    if schema.get("additionalProperties") is False:
        for name in arguments:
            if name not in properties:
                problems.append(f"unexpected argument {name!r}")

    for name, value in arguments.items():
        prop = properties.get(name)
        if not prop:
            continue
        problems.extend(_check_value(name, value, prop))

    return problems


_JSON_TYPES: dict[str, tuple[type, ...]] = {
    "string": (str,),
    "integer": (int,),
    "number": (int, float),
    "boolean": (bool,),
    "array": (list, tuple),
    "object": (dict,),
    "null": (type(None),),
}


def _check_value(name: str, value: Any, prop: dict[str, Any]) -> list[str]:
    problems: list[str] = []
    expected = prop.get("type")

    if expected in _JSON_TYPES:
        allowed = _JSON_TYPES[expected]
        # ``bool`` is a subclass of ``int``; do not let True pass as an integer.
        if expected in ("integer", "number") and isinstance(value, bool):
            problems.append(f"{name!r} must be {expected}, got boolean")
        elif not isinstance(value, allowed):
            problems.append(
                f"{name!r} must be {expected}, got {type(value).__name__}"
            )

    if "enum" in prop and value not in prop["enum"]:
        allowed_values = ", ".join(repr(v) for v in prop["enum"])
        problems.append(f"{name!r} must be one of: {allowed_values}")

    if expected == "array" and isinstance(value, (list, tuple)):
        item_schema = prop.get("items") or {}
        if item_schema:
            for index, item in enumerate(value):
                problems.extend(_check_value(f"{name}[{index}]", item, item_schema))

    return problems

File: go/mcp/schema.go
package mcp

import "fmt"

// Property describes one tool argument.
type Property struct {
	Type        string
	Description string
	Enum        []any
	Items       map[string]any
}

func (p Property) toMap() map[string]any {
	out := map[string]any{}
	if p.Type != "" {
		out["type"] = p.Type
	}
	if p.Description != "" {
		out["description"] = p.Description
	}
	if p.Enum != nil {
		out["enum"] = p.Enum
	}
	if p.Items != nil {
		out["items"] = p.Items
	}
	return out
}

// ObjectSchema builds the inputSchema object MCP expects.
//
//	mcp.ObjectSchema(map[string]mcp.Property{
//	    "order_id": {Type: "string", Description: "Internal order identifier."},
//	}, []string{"order_id"})
func ObjectSchema(properties map[string]Property, required []string) map[string]any {
	props := map[string]any{}
	for name, property := range properties {
		props[name] = property.toMap()
	}
	schema := map[string]any{
		"type":                 "object",
		"properties":           props,
		"additionalProperties": false,
	}
	if len(required) > 0 {
		schema["required"] = required
	}
	return schema
}

// ValidateArguments checks presence, primitive type, enum membership and
// unknown keys. Same deliberately small surface as the Python version: it
// catches what models get wrong without pulling in a schema library.
func ValidateArguments(schema map[string]any, arguments map[string]any) []string {
	var problems []string

	properties, _ := schema["properties"].(map[string]any)

	if required, ok := schema["required"].([]string); ok {
		for _, name := range required {
			if _, present := arguments[name]; !present {
				problems = append(problems, fmt.Sprintf("missing required argument %q", name))
			}
		}
	}

	if additional, ok := schema["additionalProperties"].(bool); ok && !additional {
		for name := range arguments {
			if _, declared := properties[name]; !declared {
				problems = append(problems, fmt.Sprintf("unexpected argument %q", name))
			}
		}
	}

	for name, value := range arguments {
		property, ok := properties[name].(map[string]any)
		if !ok {
			continue
		}
		problems = append(problems, checkValue(name, value, property)...)
	}
	return problems
}

func checkValue(name string, value any, property map[string]any) []string {
	var problems []string

	if expected, ok := property["type"].(string); ok && !matchesJSONType(value, expected) {
		problems = append(problems,
			fmt.Sprintf("%q must be %s, got %T", name, expected, value))
	}

	if allowed, ok := property["enum"].([]any); ok {
		found := false
		for _, candidate := range allowed {
			if candidate == value {
				found = true
				break
			}
		}
		if !found {
			problems = append(problems, fmt.Sprintf("%q is not one of the allowed values", name))
		}
	}

	if items, ok := property["items"].(map[string]any); ok {
		if list, isList := value.([]any); isList {
			for index, item := range list {
				problems = append(problems,
					checkValue(fmt.Sprintf("%s[%d]", name, index), item, items)...)
			}
		}
	}
	return problems
}

// matchesJSONType compares against encoding/json's decoded shapes: every
// number arrives as float64, so "integer" means a float64 with no fraction.
func matchesJSONType(value any, expected string) bool {
	switch expected {
	case "string":
		_, ok := value.(string)
		return ok
	case "boolean":
		_, ok := value.(bool)
		return ok
	case "number":
		_, ok := value.(float64)
		return ok
	case "integer":
		number, ok := value.(float64)
		return ok && number == float64(int64(number))
	case "array":
		_, ok := value.([]any)
		return ok
	case "object":
		_, ok := value.(map[string]any)
		return ok
	case "null":
		return value == nil
	default:
		return true
	}
}
